build_modules/lib/src/module_library.dart

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/analyzer.dart';
+import 'package:build/build.dart';
+import 'package:meta/meta.dart';
+
+class ModuleLibrary {
+  /// The AssetId of the original Dart source file.
+  final AssetId id;
+
+  /// Whether this library can be imported.
+  ///
+  /// This will be false if the source file is a "part of", or imports code that
+  /// can't be used outside the SDK.
+  final bool isImportable;
+
+  /// Whether this library is an entrypoint.
+  ///
+  /// True if the library is in `lib/` but not `lib/src`, or if it is outside of
+  /// `lib/` and contains a `main` method. Always false if this is not an
+  /// importable library.
+  final bool isEntryPoint;
+
+  /// The IDs of libraries that are imported or exported by this library.
+  ///
+  /// Null if this is not an importable library.
+  final Set<AssetId> deps;
+
+  /// The "part" files for this library.
+  ///
+  /// Null if this is not an importable library.
+  final Set<AssetId> parts;
+
+  ModuleLibrary._(this.id,
+      {@required this.isEntryPoint, @required this.deps, @required this.parts})
+      : isImportable = true;
+
+  ModuleLibrary._nonImportable(this.id)
+      : isImportable = false,
+        isEntryPoint = false,
+        deps = null,
+        parts = null;
+
+  factory ModuleLibrary._fromCompilationUnit(
+      AssetId id, bool isEntryPoint, CompilationUnit parsed) {
+    var deps = new Set<AssetId>();
+    var parts = new Set<AssetId>();
+    for (var directive in parsed.directives) {
+      if (directive is! UriBasedDirective) continue;
+      var path = (directive as UriBasedDirective).uri.stringValue;
+      if (Uri.parse(path).scheme == 'dart') continue;
+      var linkedId = new AssetId.resolve(path, from: id);
+      if (linkedId == null) continue;
+      if (directive is PartDirective) {
+        parts.add(linkedId);
+        continue;
+      }
+
+      List<Configuration> conditionalDirectiveConfigurations;
+
+      if (directive is ImportDirective && directive.configurations.isNotEmpty) {
+        conditionalDirectiveConfigurations = directive.configurations;
+      } else if (directive is ExportDirective &&
+          directive.configurations.isNotEmpty) {
+        conditionalDirectiveConfigurations = directive.configurations;
+      }
+      deps.add(linkedId);
+      if (conditionalDirectiveConfigurations != null) {
+        deps.addAll(conditionalDirectiveConfigurations
+            .map((c) => Uri.parse(c.uri.stringValue))
+            .where((u) => u.scheme != 'dart')
+            .map((u) => new AssetId.resolve(u.toString(), from: id)));
```

Comment:
What's going on here?

Reply:
Moved from https://github.com/dart-lang/build/pull/1649/files#diff-e6beb981c2249722d51e60e34b104f88L109

Same thing that's happening here: https://github.com/dart-lang/build/pull/1649/files#diff-08a717a90be39b7bdd847098ce7bc663R54

We'll have an import URI thats with either `package:something/something.dart`; or `relative/import.dart`, and so we take all the URIs and turn them into AssetIds (which are absolute) by resolving them against where the import comes from.